posthog/api/capture.py

```diff
@@ -464,11 +475,59 @@ def capture_internal(event, distinct_id, ip, site_url, now, sent_at, team_id, ev
     kafka_partition_key = None
 
     if event["event"] in ("$snapshot", "$performance_event"):
-        kafka_partition_key = None
-    else:
-        candidate_partition_key = f"{team_id}:{distinct_id}"
+        return log_event(parsed_event, event["event"], partition_key=kafka_partition_key)
 
-        if candidate_partition_key not in settings.EVENT_PARTITION_KEYS_TO_OVERRIDE:
-            kafka_partition_key = hashlib.sha256(candidate_partition_key.encode()).hexdigest()
+    candidate_partition_key = f"{team_id}:{distinct_id}"
+
+    if is_randomly_partitioned(candidate_partition_key) is False:
+        kafka_partition_key = hashlib.sha256(candidate_partition_key.encode()).hexdigest()
 
     return log_event(parsed_event, event["event"], partition_key=kafka_partition_key)
+
+
+def is_randomly_partitioned(candidate_partition_key: str) -> bool:
+    """Check whether event with given partition key is to be randomly partitioned.
+
+    Checking whether an event should be randomly partitioned is a two step process:
+
+    1. Using a token-bucket algorithm, check if the event's candidate key has exceeded
+       the given PARTITION_KEY_BUCKET_CAPACITY. If it has, events with that key could
+       be experiencing a temporary burst in traffic and should be randomly partitioned.
+       Otherwise, go to 2.
+
+    2. Check if the candidate partition key is set in the
+       EVENT_PARTITION_KEYS_TO_OVERRIDE instance setting. If it is, then the event
+       should be randomly partitioned. Otherwise, no random partition should occur and
+       the candidate partition key can be used.
+
+    Token-bucket algorithm (step 1) is ignored if the
+    PARTITION_KEY_AUTOMATIC_OVERRIDE_ENABLED setting is set to False.
+
+    Args:
+        candidate_partition_key: The partition key that would be used if we decide
+            on no random partitioniong. This is in the format `team_id:distinct_id`.
+
+    Returns:
+        Whether the given partition key should be used.
+    """
+    if settings.PARTITION_KEY_AUTOMATIC_OVERRIDE_ENABLED:
+
+        has_capacity = LIMITER.consume(candidate_partition_key)
+
+        if has_capacity is False:
+
+            if not LOG_RATE_LIMITER.consume(candidate_partition_key):
+                # Return early if we have logged this key already.
+                return True
+
+            statsd.incr("partition_key_capacity_exceeded", tags={"partition_key": candidate_partition_key})
+            logger.warning(
+                "Partition key %s overriden as bucket capacity of %s tokens exceeded",
+                candidate_partition_key,
+                LIMITER._capacity,
+            )
```

Comment:
```suggestion
            logger.warning(
                "Partition key %s overridden as bucket capacity of %s tokens exceeded",
                candidate_partition_key,
                LIMITER._capacity,
            )
```

non-blocking and off topic:

It might be better on the log analysis side to use a structured log here e.g. passing the log params something like 

```
logger.warning("partition_key_overridden_at_capacity", {"candidate_partition_key": candidate_partition_key, "capacity": LIMITER._capacity})
```

Then we can e.g. create histograms with LogQL etc. but that's maybe a bigger discussion around log formatting guidelines.